File: simplified-books-core-ng/src/main/java/org/nypl/simplified/books/controller/BookSyncTask.java
package org.nypl.simplified.books.controller;

import com.io7m.jfunctional.Option;
import com.io7m.jfunctional.OptionType;
import com.io7m.jfunctional.Some;
import com.io7m.jfunctional.Unit;
import com.io7m.jnull.NullCheck;

import org.nypl.simplified.books.accounts.AccountAuthenticatedHTTP;
import org.nypl.simplified.books.accounts.AccountAuthenticationCredentials;
import org.nypl.simplified.books.accounts.AccountProviderAuthenticationDescription;
import org.nypl.simplified.books.accounts.AccountType;
import org.nypl.simplified.books.book_database.Book;
import org.nypl.simplified.books.book_database.BookDatabaseEntryType;
import org.nypl.simplified.books.book_database.BookDatabaseException;
import org.nypl.simplified.books.book_database.BookDatabaseType;
import org.nypl.simplified.books.book_database.BookID;
import org.nypl.simplified.books.book_database.BookIDs;
import org.nypl.simplified.books.book_registry.BookRegistryType;
import org.nypl.simplified.books.book_registry.BookStatus;
import org.nypl.simplified.books.book_registry.BookWithStatus;
import org.nypl.simplified.http.core.HTTPAuthType;
import org.nypl.simplified.http.core.HTTPResultError;
import org.nypl.simplified.http.core.HTTPResultException;
import org.nypl.simplified.http.core.HTTPResultMatcherType;
import org.nypl.simplified.http.core.HTTPResultOKType;
import org.nypl.simplified.http.core.HTTPResultType;
import org.nypl.simplified.http.core.HTTPType;
import org.nypl.simplified.opds.core.OPDSAcquisitionFeed;
import org.nypl.simplified.opds.core.OPDSAcquisitionFeedEntry;
import org.nypl.simplified.opds.core.OPDSAvailabilityRevoked;
import org.nypl.simplified.opds.core.OPDSAvailabilityType;
import org.nypl.simplified.opds.core.OPDSFeedParserType;
import org.nypl.simplified.opds.core.OPDSParseException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.concurrent.Callable;

final class BookSyncTask implements Callable<Unit> {

  private static final Logger LOG = LoggerFactory.getLogger(BookSyncTask.class);

  private final BooksControllerType books_controller;
  private final AccountType account;
  private final BookRegistryType book_registry;
  private final HTTPType http;
  private final OPDSFeedParserType feed_parser;

  BookSyncTask(
      final BooksControllerType books_controller,
      final AccountType account,
      final BookRegistryType book_registry,
      final HTTPType http,
      final OPDSFeedParserType feed_parser) {

    this.books_controller =
        NullCheck.notNull(books_controller, "Books controller");
    this.account =
        NullCheck.notNull(account, "Account");
    this.book_registry =
        NullCheck.notNull(book_registry, "Book registry");
    this.http =
        NullCheck.notNull(http, "Http");
    this.feed_parser =
        NullCheck.notNull(feed_parser, "Feed parser");
  }

  @Override
  public Unit call() throws Exception {
    try {
      LOG.debug("syncing account {}", this.account.id().id());
      return execute();
    } finally {
      LOG.debug("finished syncing account {}", this.account.id().id());
    }
  }

  private Unit execute() throws Exception {
    final OptionType<AccountProviderAuthenticationDescription> provider_auth_opt =
        this.account.provider().authentication();

    if (!provider_auth_opt.isSome()) {
      LOG.debug("account does not support syncing");
      return Unit.unit();
    }

    final AccountProviderAuthenticationDescription provider_auth =
        ((Some<AccountProviderAuthenticationDescription>) provider_auth_opt).get();

    final OptionType<AccountAuthenticationCredentials> credentials_opt =
        this.account.credentials();

    if (credentials_opt.isNone()) {
      LOG.debug("no credentials, aborting!");
      return Unit.unit();
    }

    final AccountAuthenticationCredentials credentials =
        ((Some<AccountAuthenticationCredentials>) credentials_opt).get();
    final HTTPAuthType auth =
        AccountAuthenticatedHTTP.createAuthenticatedHTTP(credentials);

    final HTTPResultType<InputStream> result =
        this.http.get(Option.some(auth), provider_auth.loginURI(), 0L);

    return result.matchResult(
        new HTTPResultMatcherType<InputStream, Unit, Exception>() {
          @Override
          public Unit onHTTPError(final HTTPResultError<InputStream> e) throws Exception {
            return BookSyncTask.this.onHTTPError(e, provider_auth);
          }

          @Override
          public Unit onHTTPException(final HTTPResultException<InputStream> e) throws Exception {
            throw e.getError();
          }

          @Override
          public Unit onHTTPOK(final HTTPResultOKType<InputStream> e) throws Exception {
            return BookSyncTask.this.onHTTPOK(e, provider_auth);
          }
        });
  }

  private Unit onHTTPOK(
      final HTTPResultOKType<InputStream> result,
      final AccountProviderAuthenticationDescription provider_auth) throws IOException {
    try {
      parseFeed(result, provider_auth);
      return Unit.unit();
    } finally {
      result.close();
    }
  }

  private void parseFeed(
      final HTTPResultOKType<InputStream> result,
      final AccountProviderAuthenticationDescription provider_auth)
      throws OPDSParseException {

    final OPDSAcquisitionFeed feed =
        this.feed_parser.parse(provider_auth.loginURI(), result.getValue());

    /*
     * Obtain the set of books that are on disk already. If any
     * of these books are not in the received feed, then they have
     * expired and should be deleted.
     */

    final BookDatabaseType book_database = this.account.bookDatabase();
    final Set<BookID> existing = book_database.books();

    /*
     * Handle each book in the received feed.
     */

    final Set<BookID> received = new HashSet<>(64);
    final List<OPDSAcquisitionFeedEntry> entries = feed.getFeedEntries();
    for (final OPDSAcquisitionFeedEntry opds_entry : entries) {
      final BookID book_id = BookIDs.newFromOPDSEntry(opds_entry);
      received.add(book_id);
      LOG.debug("[{}] updating", book_id.brief());

      try {
        final BookDatabaseEntryType database_entry =
            book_database.createOrUpdate(book_id, opds_entry);
        final Book book = database_entry.getBook();
        this.book_registry.update(BookWithStatus.create(book, BookStatus.fromBook(book)));
      } catch (final BookDatabaseException e) {
        LOG.error("[{}] unable to update database entry: ", book_id.brief(), e);
      }
    }

    /*
     * Now delete any book that previously existed, but is not in the
     * received set. Queue any revoked books for completion and then
     * deletion.
     */

    final Set<BookID> revoking = new HashSet<BookID>(existing.size());
    for (final BookID existing_id : existing) {
      try {
        LOG.debug("[{}] checking for deletion", existing_id.brief());

        if (!received.contains(existing_id)) {
          final BookDatabaseEntryType db_entry = book_database.entry(existing_id);
          final OPDSAvailabilityType a = db_entry.getBook().getEntry().getAvailability();
          if (a instanceof OPDSAvailabilityRevoked) {
            revoking.add(existing_id);
          }

          LOG.debug("[{}] deleting", existing_id.brief());
          db_entry.delete();
          this.book_registry.clearFor(existing_id);
        } else {
          LOG.debug("[{}] keeping", existing_id.brief());
        }
      } catch (final Throwable x) {
        LOG.error("[{}]: unable to delete entry: ", existing_id.value(), x);
      }
    }

    /*
     * Finish the revocation of any books that need it.
     */

    for (final BookID revoke_id : revoking) {
      LOG.debug("[{}] revoking", revoke_id.brief());
      this.books_controller.bookRevoke(this.account, revoke_id);
    }
  }

  private Unit onHTTPError(
      final HTTPResultError<InputStream> result,
      final AccountProviderAuthenticationDescription provider_auth) throws Exception {

    final String message =
        String.format("%s: %d: %s", provider_auth.loginURI(), result.getStatus(), result.getMessage());

    switch (result.getStatus()) {
      case HttpURLConnection.HTTP_UNAUTHORIZED: {
        this.account.setCredentials(Option.none());
        return Unit.unit();
      }
      default: {
        throw new IOException(message);
      }
    }
  }
}

File: simplified-books-controller/src/main/java/org/nypl/simplified/books/controller/BookSyncTask.kt
package org.nypl.simplified.books.controller

import com.io7m.jfunctional.Some
import org.librarysimplified.http.api.LSHTTPClientType
import org.librarysimplified.http.api.LSHTTPResponseStatus
import org.nypl.simplified.accounts.api.AccountAuthenticatedHTTP
import org.nypl.simplified.accounts.api.AccountAuthenticationCredentials
import org.nypl.simplified.accounts.api.AccountID
import org.nypl.simplified.accounts.api.AccountLoginState
import org.nypl.simplified.accounts.api.AccountProviderAuthenticationDescription
import org.nypl.simplified.accounts.api.AccountProviderType
import org.nypl.simplified.accounts.database.api.AccountType
import org.nypl.simplified.accounts.registry.api.AccountProviderRegistryType
import org.nypl.simplified.books.api.BookID
import org.nypl.simplified.books.api.BookIDs
import org.nypl.simplified.books.book_database.api.BookDatabaseEntryType
import org.nypl.simplified.books.book_database.api.BookDatabaseException
import org.nypl.simplified.books.book_registry.BookRegistryType
import org.nypl.simplified.books.book_registry.BookStatus
import org.nypl.simplified.books.book_registry.BookWithStatus
import org.nypl.simplified.books.controller.api.BooksControllerType
import org.nypl.simplified.feeds.api.FeedLoaderType
import org.nypl.simplified.feeds.api.FeedLoading
import org.nypl.simplified.opds.core.OPDSAvailabilityRevoked
import org.nypl.simplified.opds.core.OPDSFeedParserType
import org.nypl.simplified.opds.core.OPDSParseException
import org.nypl.simplified.patron.api.PatronUserProfile
import org.nypl.simplified.patron.api.PatronUserProfileParsersType
import org.nypl.simplified.profiles.api.ProfileID
import org.nypl.simplified.profiles.api.ProfilesDatabaseType
import org.nypl.simplified.taskrecorder.api.TaskRecorder
import org.nypl.simplified.taskrecorder.api.TaskResult
import org.slf4j.LoggerFactory
import java.io.ByteArrayInputStream
import java.io.IOException
import java.io.InputStream
import java.net.URI
import java.util.HashSet
import java.util.concurrent.TimeUnit

class BookSyncTask(
  private val accountID: AccountID,
  profileID: ProfileID,
  profiles: ProfilesDatabaseType,
  private val booksController: BooksControllerType,
  private val accountRegistry: AccountProviderRegistryType,
  private val bookRegistry: BookRegistryType,
  private val feedLoader: FeedLoaderType,
  private val patronParsers: PatronUserProfileParsersType,
  private val http: LSHTTPClientType,
  private val feedParser: OPDSFeedParserType
) : AbstractBookTask(accountID, profileID, profiles) {

  override val logger =
    LoggerFactory.getLogger(BookSyncTask::class.java)

  override val taskRecorder =
    TaskRecorder.create()

  override fun execute(account: AccountType): TaskResult.Success<Unit> {
    this.logger.debug("syncing account {}", account.id)
    this.taskRecorder.beginNewStep("Syncing...")

    val provider = this.updateAccountProvider(account)
    val providerAuth = provider.authentication
    if (providerAuth == AccountProviderAuthenticationDescription.Anonymous) {
      this.logger.debug("account does not support syncing")
      return this.taskRecorder.finishSuccess(Unit)
    }

    val credentials = account.loginState.credentials
    if (credentials == null) {
      this.logger.debug("no credentials, aborting!")
      return this.taskRecorder.finishSuccess(Unit)
    }

    this.fetchPatronUserProfile(
      account = account,
      credentials = credentials
    )

    val loansURI = provider.loansURI
    if (loansURI == null) {
      this.logger.debug("no loans URI, aborting!")
      return this.taskRecorder.finishSuccess(Unit)
    }

    val request =
      this.http.newRequest(loansURI)
        .setAuthorization(AccountAuthenticatedHTTP.createAuthorization(credentials))
        .build()

    val response = request.execute()
    return when (val status = response.status) {
      is LSHTTPResponseStatus.Responded.OK -> {
        this.onHTTPOK(status.bodyStream ?: ByteArrayInputStream(ByteArray(0)), provider, account)
        this.taskRecorder.finishSuccess(Unit)
      }
      is LSHTTPResponseStatus.Responded.Error -> {
        val recovered = this.onHTTPError(status, account)

        if (recovered) {
          this.taskRecorder.finishSuccess(Unit)
        } else {
          val message = String.format("%s: %d: %s", provider.loansURI, status.properties.status, status.properties.message)
          val exception = IOException(message)
          this.taskRecorder.currentStepFailed(
            message = message,
            errorCode = "syncFailed",
            exception = exception
          )
          throw TaskFailedHandled(exception)
        }
      }
      is LSHTTPResponseStatus.Failed ->
        throw IOException(status.exception)
    }
  }

  private fun fetchPatronUserProfile(
    account: AccountType,
    credentials: AccountAuthenticationCredentials
  ) {
    try {
      val profile =
        PatronUserProfiles.runPatronProfileRequest(
          taskRecorder = this.taskRecorder,
          patronParsers = this.patronParsers,
          credentials = credentials,
          http = this.http,
          account = account
        )

      account.updateCredentialsIfAvailable {
        this.withNewAnnotationsURI(it, profile)
      }
    } catch (e: Exception) {
      this.logger.error("patron user profile: ", e)
    }
  }

  private fun withNewAnnotationsURI(
    currentCredentials: AccountAuthenticationCredentials,
    profile: PatronUserProfile
  ): AccountAuthenticationCredentials {
    return when (currentCredentials) {
      is AccountAuthenticationCredentials.Basic ->
        currentCredentials.copy(annotationsURI = profile.annotationsURI)
      is AccountAuthenticationCredentials.OAuthWithIntermediary ->
        currentCredentials.copy(annotationsURI = profile.annotationsURI)
      is AccountAuthenticationCredentials.SAML2_0 ->
        currentCredentials.copy(annotationsURI = profile.annotationsURI)
    }
  }

  override fun onFailure(result: TaskResult.Failure<Unit>) {
    // Nothing to do
  }

  private fun updateAccountProvider(account: AccountType): AccountProviderType {
    this.logger.debug("resolving the existing account provider")

    val oldProvider = account.provider
    var newDescription =
      this.accountRegistry.findAccountProviderDescription(oldProvider.id)
    if (newDescription == null) {
      this.logger.debug("could not find account description for {} in registry", oldProvider.id)
      newDescription = oldProvider.toDescription()
    } else {
      this.logger.debug("found account description for {} in registry", oldProvider.id)
    }

    val newProviderResult =
      this.accountRegistry.resolve(
        { accountProvider, message ->
          this.logger.debug("[{}]: {}", accountProvider, message)
        },
        newDescription
      )

    return when (newProviderResult) {
      is TaskResult.Success -> {
        this.logger.debug("successfully resolved the account provider")
        account.setAccountProvider(newProviderResult.result)
        newProviderResult.result
      }
      is TaskResult.Failure -> {
        this.logger.error("failed to resolve account provider: ", newProviderResult.exception)
        oldProvider
      }
    }
  }

  @Throws(IOException::class)
  private fun onHTTPOK(
    stream: InputStream,
    provider: AccountProviderType,
    account: AccountType
  ) {
    return stream.use { ok ->
      this.parseFeed(ok, provider, account)
    }
  }

  @Throws(OPDSParseException::class)
  private fun parseFeed(
    stream: InputStream,
    provider: AccountProviderType,
    account: AccountType
  ) {
    val feed = this.feedParser.parse(provider.loansURI, stream)

    /*
     * Obtain the set of books that are on disk already. If any
     * of these books are not in the received feed, then they have
     * expired and should be deleted.
     */

    val bookDatabase = account.bookDatabase
    val existing = bookDatabase.books()

    /*
     * Handle each book in the received feed.
     */

    val received = HashSet<BookID>(64)
    val entries = feed.feedEntries
    for (opdsEntry in entries) {
      val bookId = BookIDs.newFromOPDSEntry(opdsEntry)
      received.add(bookId)
      this.logger.debug("[{}] updating", bookId.brief())

      try {
        val databaseEntry = bookDatabase.createOrUpdate(bookId, opdsEntry)
        val book = databaseEntry.book
        this.bookRegistry.update(BookWithStatus(book, BookStatus.fromBook(book)))
      } catch (e: BookDatabaseException) {
        this.logger.error("[{}] unable to update database entry: ", bookId.brief(), e)
      }
    }

    /*
     * Now delete/revoke any book that previously existed, but is not in the
     * received set.
     */

    val revoking = HashSet<BookID>(existing.size)
    for (existingId in existing) {
      try {
        this.logger.debug("[{}] checking for deletion", existingId.brief())

        if (!received.contains(existingId)) {
          val dbEntry = bookDatabase.entry(existingId)
          val a = dbEntry.book.entry.availability
          if (a is OPDSAvailabilityRevoked) {
            revoking.add(existingId)
          } else {
            this.logger.debug("[{}] deleting", existingId.brief())
            this.updateRegistryForBook(account, dbEntry)
            dbEntry.delete()
          }
        } else {
          this.logger.debug("[{}] keeping", existingId.brief())
        }
      } catch (x: Throwable) {
        this.logger.error("[{}]: unable to delete entry: ", existingId.value(), x)
      }
    }

    /*
     * Finish the revocation of any books that need it.
     */

    for (revoke_id in revoking) {
      this.logger.debug("[{}] revoking", revoke_id.brief())
      this.booksController.bookRevoke(account.id, revoke_id)
    }
  }

  private fun updateRegistryForBook(
    account: AccountType,
    dbEntry: BookDatabaseEntryType
  ) {
    this.logger.debug("attempting to fetch book permalink to update registry")
    val alternateOpt = dbEntry.book.entry.alternate
    return if (alternateOpt is Some<URI>) {
      val alternate = alternateOpt.get()
      val entry =
        FeedLoading.loadSingleEntryFeed(
          feedLoader = this.feedLoader,
          taskRecorder = this.taskRecorder,
          accountID = this.accountID,
          uri = alternate,
          timeout = Pair(30L, TimeUnit.SECONDS),
          httpAuth = null,
          method = "GET"
        )

      /*
       * Write a new book database entry based on the server state, and pretend that all the
       * books have been deleted. The code will delete the entire database entry after this
       * method returns anyway, so this code ensures that something sensible goes into the
       * book registry.
       */

      dbEntry.writeOPDSEntry(entry.feedEntry)
      val newBook = dbEntry.book.copy(formats = emptyList())
      val status = BookStatus.fromBook(newBook)

      this.logger.debug("book's new state is {}", status)
      this.bookRegistry.update(BookWithStatus(newBook, status))
    } else {
      throw IOException("No alternate link is available")
    }
  }

  /**
   * Returns whether we recovered from the error.
   */

  private fun onHTTPError(
    result: LSHTTPResponseStatus.Responded.Error,
    account: AccountType
  ): Boolean {
    if (result.properties.status == 401) {
      this.logger.debug("removing credentials due to 401 server response")
      account.setLoginState(AccountLoginState.AccountNotLoggedIn)
      return true
    }
    return false
  }
}
